---
title: "qDNASeq"
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# setwd("~/Documents/Projects/cancer/sWGS")
library(bedr)
library(biomaRt)
library(dplyr)
library(ggplot2)
library(ggfortify)
library("ggpubr")
library(plotly) 
library(tidyr)
library(knitr)
library(reshape2)
```

# Introduction

This report investigates the output from qDNASeq and compare the output to WGS of the same sample from GEL. The study design is there are 8 patiensts, each with 2 FF sample and 2 FFPE samples, giving us a total of 32 samples. The samples have been processed on the NovaSeq machine using shallow WGS technique, therefore we expect very low coverages. The first part will look at general QC of the data, the second part will look at each sample in detail at different coverages (normal vs downsampled) and different bins, with comparison to WGS and concordance between tissue types.




- Also add extra column with info on FF and FFPE
- Compare normal and downsampled at each bin
  - Plot: x=bin size, y=TP,FP,FN (have normal and downsampled in one plot (shading & color))
- Similiarity between FF and FFPE
- Case 5 PCA plot

Section:
Case 5:
  - Per bin
    - Summary of truth with Y/N ff ffpe
    - Summary FP,FN and TP table
    - Performance of FFPE and FF 
      - How TP for FFPE and FF from the truth table
  - Effect of bin size and downsamples with plot (2 plot for tissue type)
  

``` {r functions}
# function to clean qDNASeq summarised data
make_bed_from_sum_df <- function(SWGS_sum) {
  # make loc column
  SWGS_sum$location <- paste(SWGS_sum$CHROMOSOME, SWGS_sum$START, sep=":")
  SWGS_sum$location <- paste(SWGS_sum$location, SWGS_sum$STOP, sep="-")
  # order and make bed file
  SWGS_sum <- as.data.frame(SWGS_sum$location)
  SWGS_sum <- as.data.frame(SWGS_sum[order(SWGS_sum$`SWGS_sum$location`),])
  # remove dup rows
  SWGS_sum <- as.data.frame(SWGS_sum[!duplicated(SWGS_sum), ])
  colnames(SWGS_sum) <- "location"
  SWGS_sum <- separate(data = SWGS_sum, col = location, into = c("chr", "start", "end"))

  return(SWGS_sum)
  
}

bed_intersect <- function(sample_bed, truth_bed){
  # both dataframes must be chr,start,end
  # truth is expected to have X or Y chr 
  # pred is not expected to have X or Y 
  
  # finds the interaction of the sample and truth bed file
  
  # convert truth data to int, do not touch chr
  truth_bed$start <- as.integer(truth_bed$start)
  truth_bed$end <- as.integer(truth_bed$end)

  # convert pred data to int
  sample_bed$chr <- as.character(sample_bed$chr)
  sample_bed$start <- as.integer(sample_bed$start)
  sample_bed$end <- as.integer(sample_bed$end)
  
  # sort the bed files
  truth_bed.sort <- bedr.sort.region(truth_bed, check.chr = F, verbose = F)
  sample_bed.sort <- bedr.sort.region(sample_bed, check.chr = F, verbose = F)
  
  # remove dup rows
  truth_bed.sort <- truth_bed.sort[!duplicated(truth_bed.sort), ] 
  sample_bed.sort <- sample_bed.sort[!duplicated(sample_bed.sort), ] 
  # intersect
  pred.int.truth <- bedr(
                      input = list(a = sample_bed.sort, b = truth_bed.sort),
                      method = "intersect",
                      params = "-loj -sorted",
                      check.chr = F, verbose = F, engine = "bedtools", check.valid = T)
  return(pred.int.truth)
  
}

bed_missed_in_sample <- function(sample_bed, truth_bed){
  # both dataframes must be chr,start,end
  # truth is expected to have X or Y chr 
  # pred is not expected to have X or Y 
  
  # finds the interaction of the sample and truth bed file
  
  # convert truth data to int, do not touch chr
  truth_bed$start <- as.integer(truth_bed$start)
  truth_bed$end <- as.integer(truth_bed$end)

  # convert pred data to int
  sample_bed$chr <- as.character(sample_bed$chr)
  sample_bed$start <- as.integer(sample_bed$start)
  sample_bed$end <- as.integer(sample_bed$end)
  
  # sort the bed files
  truth_bed.sort <- bedr.sort.region(truth_bed, check.chr = F, verbose = F)
  sample_bed.sort <- bedr.sort.region(sample_bed, check.chr = F, verbose = F)
  
  # remove dup rows
  truth_bed.sort <- truth_bed.sort[!duplicated(truth_bed.sort), ] 
  sample_bed.sort <- sample_bed.sort[!duplicated(sample_bed.sort), ] 
  # intersect
  pred.int.truth <- bedr(
                      input = list(a =truth_bed.sort , b = sample_bed.sort),
                      method = "intersect",
                      params = "-loj -sorted",
                      check.chr = F, verbose = F, engine = "bedtools", check.valid = T)
  return(pred.int.truth)
  
}

######### Big function which will:
# - calculate TP,FP and FN
# - Add sample column in the truth wgs with yes if there is overlap with sWGS
# - Make separate tables for TP, FP and FN with gene annotation

sWGS_vs_WGS <- function(Case_WGS, metadata, sample, mydir, subfolder, attributes, filters, mart){
  # get the summaried data for samples associated to this case
  C_samples <- unique(metadata[which(metadata$CUH_number %in% sample), "samplename"])
  C_fullsamplenames <-  unique(metadata[which(metadata$CUH_number %in% sample), "fullsamplename"])
  Case_truthvspredicted_df <- as.data.frame(matrix(NA, nrow =4,ncol=3))
  colnames(Case_truthvspredicted_df) <- c("TP", "FP", "TN")
  rownames(Case_truthvspredicted_df) <- C_fullsamplenames
  Case_WGS[, C_fullsamplenames] <- NA
  # make a bed file output to do intersect with these 4 samples
  print(paste("Truth WGS", sample, "has CN segements = ", nrow(Case_WGS)))
  for(i in 1:length(C_samples)) { 
    N = C_samples[i]
    N_full =  C_fullsamplenames[i]
    # assign the new samplename with  _bed to the new datafraame that contains loc,start,end
    assign(paste(N, "_bed", sep=""), make_bed_from_sum_df(get(N)))
    print(paste(N, " has CN segements = ", nrow( make_bed_from_sum_df(get(N))), sep = ""))
    # save the dataframe
    write.table(get(paste(N, "_bed", sep="")), file = paste(mydir, "/", subfolder, "/", sample, "_" , N, "_predicted_CN_locs.tsv", sep=""), sep = "\t",
                col.names = F, row.names = F, quote = F)
    
    # compare to the truth dataset
    # contains all intersect with TP and FP
    intersect_loj <- bed_intersect(sample_bed = get(paste(N, "_bed", sep="")), truth_bed = Case_WGS_location_bed)
    # row = sample, column = TP
    # if true, all will be coordinates
    TP <- intersect_loj[which(intersect_loj[,4] > 0),]
    Case_truthvspredicted_df[i,1] <- nrow(TP)
    # if true, write yes as to whether this yes is for the FF or the FFPE tissue
    # first get the truth location as chr:start-end
    TP$truth_location <- paste(TP$chr.b, ":", TP$start.b, "-", TP$end.b, sep = "")
    TP$yes <- rep("yes", nrow(TP))
    # if the Case location and the TP truth location match, then say past yes in the sample tissue column
    Case_WGS[,N_full] <- TP$yes[match(Case_WGS$Position..hg38.,TP$truth_location)]
    # row = sample, column = FP
    # if FP, they will exist in 4,5,6 column (truth dataset) as -1
    Case_truthvspredicted_df[i,2] <- nrow(intersect_loj[which(intersect_loj[,4] == "."),])
    # row = sample, column = TN
    TN <- bed_missed_in_sample(sample_bed =  get(paste(N, "_bed", sep="")), truth_bed = Case_WGS_location_bed)
    Case_truthvspredicted_df[i,3] <- nrow(TN[which(TN[,4] == "."),])
    
    ## Output all the regions + gene annotation
    # select sample
    sample_sum_file <- get(N)
    sample_sum_file$location <- paste(sample_sum_file$CHROMOSOME, ":",sample_sum_file$START, "-",sample_sum_file$STOP, sep = "")
    all.genes.df <- as.data.frame(matrix(NA, nrow = 0, ncol = 6))
    colnames(all.genes.df) <- c( "ensembl_gene_id","hgnc_symbol","chromosome_name","start_position","end_position","location")
    for(i in 1:nrow(sample_sum_file)) {
      # tell all the locations we want
      values <- list(chromosome=sample_sum_file[i,"CHROMOSOME"],start=sample_sum_file[i,"START"],end=sample_sum_file[i,"STOP"])
      # get all genes in the chr and start/end locations
      all.genes <- getBM(attributes=attributes, filters=filters, values=values, mart=mart)
       if (dim(all.genes)[1] == 0) {
        next
       }
      # the location will be what CN region contins these genes
      all.genes$location <- rep(sample_sum_file[i,"location"], nrow(all.genes))
      # some regions are not genes but novel transcripts by ensembl, so can remove these
      all.genes <- all.genes[which(all.genes$hgnc_symbol != ""),]
      all.genes.df <- rbind(all.genes.df, all.genes)
    }
    # 
    sample_sum_file <- left_join(sample_sum_file, all.genes.df, by="location")
    
    write.table(sample_sum_file,
                file = paste(mydir, subfolder, "/", sample, "_" , N, "_predicted_CN_gene_ann.tsv", sep=""),
                sep = "\t", col.names = T, row.names = F, quote = F)
    }
  
  kable(Case_truthvspredicted_df, caption = paste(sample, "vs truth WGS dataset"))
  write.table(Case_truthvspredicted_df, 
              file = paste(mydir,"/", subfolder, "/", sample ,"_TP_FP_TN.tsv", sep =""),
              sep = "\t", col.names = T, row.names = T, quote = F)
  
  Case_WGS_filtered <- Case_WGS[,c("Gene.Transcript", "Copy.Number..MCC.", "Cytogenetic.band", "Position..hg38.",
                                   "Log2.FC", "SWGS_N4_3134 _ FF", "SWGS_N20_3134R _ FF", "SWGS_N12_3016 _ FFPE",  "SWGS_N28_3016R _ FFPE")]
  write.table(Case_WGS_filtered, 
              file = paste(mydir,"/", subfolder, "/", sample ,"_TP_WGS_FF_vs_FFPE.tsv", sep =""),
              sep = "\t", col.names =T, row.names = F, quote = F)
  
  head(Case_WGS_filtered)
  
}
```

# Quality control

General QC can be found in the multiQC report on DNAnexus at: https://platform.dnanexus.com/projects/GQXY5k84926v43v88yY1FX5z/data/multiqc 
In this report we will look at the QualiMap coverages and the PCA plots

## QualiMap

sWGS targets to have low mapped reads, 5X should be sufficient to do CNV calling on. We can see in the boxplot below that the distribution is quiet varied with the mean coverage for this being 9X, higest is 21X and lowest is 0X

```{r qualimap}
qualimap <- read.csv("../qualimap_coverage.tsv", sep = "\t")
p <- ggplot(qualimap, aes(x=mean.coverageData)) +
  geom_boxplot() +
  geom_dotplot(stackdir='center', dotsize=0.30) +
  coord_flip() +
  theme_classic() +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank())

p

```

## PCA plot

The PCA plot will indicate to us what the variation in our data is driven by. This will indicate any batch effects we should be aware of and take into consideration for downstream analysis. The data we will use in the raw data as it has the same bins applied and therefore the comparison is more useful than segment data where there is little overlap across all samples.

What we find is that variation is not driven by tissue difference, this means that the log2 mean ratio calculations is not associated with tissue type. This will allow to compare the data as normal and that any variation in data is not due to tissue type. We also wanted to see whether coverage drove any differences between the qDNASeq calculations and we can see that it didn't explain as there are samples near each other have having coverage calculation in either side of the coverage distribution for this sample.

```{r read_data}
# read  data in
mydir = "~/Documents/Projects/cancer/sWGS/"

### raw data
total_files =  list.files(path = paste0(mydir), pattern= ("*.raw_calls_all_depths.tsv"))
raw_df  <- as.data.frame(matrix(ncol = 0, nrow = 51019))

for(i in 1:length(total_files)) { #up to 4 samples
  file <- total_files[i]
  file_name <- strsplit(file, ".raw_calls_all_depths.tsv") #removes the .tsv taken from file name
  tsv_file <- read.csv(paste(mydir, "/", file, sep= ""), sep="\t", header = T) #read in the file as a dataframe
  raw_df <- cbind(raw_df, tsv_file)
  file_name <- file_name[[1]] #get the dataframe from the list
  assign(file_name, tsv_file) #assigns the tsv_file to the object file_name which is the dataframe
  file_name <- NULL
  tsv_file <- NULL
}

metadata <- read.delim(paste(mydir, "metadata.txt", sep= "/"), sep = "\t")
metadata$fullsamplename <- paste(metadata$samplename,"_",metadata$Tissue.Type)

# read in biomart info 
mart <- useMart("ensembl")
mart <- useDataset("hsapiens_gene_ensembl", mart)
attributes <- c("ensembl_gene_id", "hgnc_symbol", "chromosome_name", "start_position", "end_position")
filters <- c("chromosome_name","start","end")

```

```{r PCA_plot_all}
# transform the data to be readt for prcomp
raw_df_2 <- raw_df[ , grepl( "SWGS" , names( raw_df ) ) ]
traw_df_2 <- as.data.frame(t(raw_df_2))
colnames(traw_df_2) <- raw_df_2$feature
traw_df_2$samples <- gsub("_S.*","",rownames(traw_df_2))
traw_df_2$tissuetype <- metadata$Tissue.Type[match(metadata$samplename, traw_df_2$samples)]

## PCA plot
logration_df_pca <- prcomp(traw_df_2[,c(1:51019)],
                   center = TRUE)
logration_df_pca.plot <- autoplot(logration_df_pca,
                          data = traw_df_2,
                          colour = 'samples',
                          shape = 'tissuetype',
                          main = 'PCA plot using raw data')
logration_df_pca.plot
ggplotly(logration_df_pca.plot)

```


```{r}
# lets remove the "bad QC samples"
row2remove <- c("SWGS_N7_7096_S7_L001_markdup", "SWGS_N32_7276R_S32_L001_markdup", "SWGS_N11_1911_S11_L001_markdup")

traw_df_3 <- traw_df_2[!(row.names(traw_df_2) %in% row2remove), ]
logration_df_pca <- prcomp(traw_df_3[,c(1:51019)],
                   center = TRUE)
logration_df_pca.plot <- autoplot(logration_df_pca,
                          data = traw_df_3,
                          colour = 'samples',
                          shape = 'tissuetype',
                          main = 'PCA plot using raw data (removal of outliers N7, N32, N11)')
logration_df_pca.plot
ggplotly(logration_df_pca.plot)

# can try to plot using coverage os estimating point
traw_df_3$coverage <- qualimap$mean.coverageData[match(rownames(traw_df_3), qualimap$Sample)]
logration_df_pca.plot <- autoplot(logration_df_pca,
                          data = traw_df_3,
                          colour = 'coverage',
                          shape = 'tissuetype',
                          main = 'PCA plot using raw data (removal of outliers N7, N32, N11)')
logration_df_pca.plot
ggplotly(logration_df_pca.plot)
```

# Case analysis


## Case 4 

### Normal Bin = 50kbp


```{r}
sample="CUH189"
# read  data in
mydir = "~/Documents/Projects/cancer/sWGS/"
# read in truth data
Case_WGS <- read.csv(paste(mydir, paste(sample, "CNVs.csv"), sep =""))

# output just the bed file
# remove chr 
Case_WGS$Position..hg38.<-gsub("chr","",as.character(Case_WGS$Position..hg38.))
Case_WGS_location <- separate(data = Case_WGS, col = Position..hg38., into = c("chr", "start", "end"))
Case_WGS_location_bed <- Case_WGS_location[,c("chr", "start", "end")]

# save the three columns
write.table(Case_WGS_location_bed, 
            file = paste(mydir, sample ," _targeted_CN_locs.tsv", sep =""),
            sep = "\t", col.names = F, row.names = F, quote = F)
```


#### How many TP, TN and FP when comparing our predicted CNs to the truth dataset

There will a tsv file that contains the truth bed and whether each segment in the WGS exists in the sWGS, this will be denotated with a "yes" in the column. If blank, it means that it is not called in the sWGS. Also, a table will be outputted detailed the TP, FP and TN.

* TP = Predicted CN in truth dataset
* FN = Not predicted but in truth set (e.g chrX and chrY as qDNASeq does autosomal chromosomes)
* FP = Predicted but not in truth set (keep in mind that it can be germline as we dont have control)
* TN = Not predicted in CN and not in truth dataset (can't be demonstrated)

```{r}
# read in samples
subfolder = ""
total_files =  list.files(path = paste0(mydir,subfolder, sep= "" ), pattern= ("*.summarised_calls.tsv"))
sum_df  <- as.data.frame(matrix(ncol = 6, nrow = 0))
for(i in 1:length(total_files)) { #up to 4 samples
  file <- total_files[i]
  file_name <- strsplit(file, ".summarised_calls.tsv") #removes the .tsv taken from file name
  tsv_file <- read.csv(paste(mydir, subfolder, "/", file, sep= ""), sep="\t", header = T) #read in the file as a dataframe
  sum_df <- rbind(sum_df, tsv_file)
  file_name <- file_name[[1]] #get the dataframe from the list
  file_name <- gsub("_S.*", "", file_name)
  assign(file_name, tsv_file) #assigns the tsv_file to the object file_name which is the dataframe
  file_name <- NULL
  tsv_file <- NULL
}
# get the TP, FN, FP & gene annotations
sWGS_vs_WGS(Case_WGS, metadata, sample, mydir, subfolder, attributes, filters, mart)

```
### Downsampled 5kbp

#### How many TP, TN and FP when comparing our predicted CNs to the truth dataset

There will a tsv file that contains the truth bed and whether each segment in the WGS exists in the sWGS, this will be denotated with a "yes" in the column. If blank, it means that it is not called in the sWGS. Also, a table will be outputted detailed the TP, FP and TN.

* TP = Predicted CN in truth dataset
* FN = Not predicted but in truth set (e.g chrX and chrY as qDNASeq does autosomal chromosomes)
* FP = Predicted but not in truth set (keep in mind that it can be germline as we dont have control)
* TN = Not predicted in CN and not in truth dataset (can't be demonstrated)


```{r}
sample="CUH189"
# read  data in
mydir = "~/Documents/Projects/cancer/sWGS/"
subfolder = "case4_down_5bin"

# read in truth data
Case_WGS <- read.csv(paste(mydir, paste(sample, "CNVs.csv"), sep =""))

# output just the bed file
# remove chr 
Case_WGS$Position..hg38.<-gsub("chr","",as.character(Case_WGS$Position..hg38.))
Case_WGS_location <- separate(data = Case_WGS, col = Position..hg38., into = c("chr", "start", "end"))
Case_WGS_location_bed <- Case_WGS_location[,c("chr", "start", "end")]

# read in samples
total_files =  list.files(path = paste0(mydir,subfolder, sep= "" ), pattern= ("*.summarised_calls.tsv"))
sum_df  <- as.data.frame(matrix(ncol = 6, nrow = 0))
for(i in 1:length(total_files)) { #up to 4 samples
  file <- total_files[i]
  file_name <- strsplit(file, ".summarised_calls.tsv") #removes the .tsv taken from file name
  tsv_file <- read.csv(paste(mydir, subfolder, "/", file, sep= ""), sep="\t", header = T) #read in the file as a dataframe
  sum_df <- rbind(sum_df, tsv_file)
  file_name <- file_name[[1]] #get the dataframe from the list
  file_name <- gsub("_S.*", "", file_name)
  assign(file_name, tsv_file) #assigns the tsv_file to the object file_name which is the dataframe
  file_name <- NULL
  tsv_file <- NULL
}
sWGS_vs_WGS(Case_WGS, metadata, sample, mydir, subfolder, attributes, filters, mart)
```


### Downsampled 15kbp

#### How many TP, TN and FP when comparing our predicted CNs to the truth dataset

There will a tsv file that contains the truth bed and whether each segment in the WGS exists in the sWGS, this will be denotated with a "yes" in the column. If blank, it means that it is not called in the sWGS. Also, a table will be outputted detailed the TP, FP and TN.

* TP = Predicted CN in truth dataset
* FN = Not predicted but in truth set (e.g chrX and chrY as qDNASeq does autosomal chromosomes)
* FP = Predicted but not in truth set (keep in mind that it can be germline as we dont have control)
* TN = Not predicted in CN and not in truth dataset (can't be demonstrated)


```{r}
sample="CUH189"
# read  data in
mydir = "~/Documents/Projects/cancer/sWGS/"
subfolder = "case4_down_15bin"

# read in truth data
Case_WGS <- read.csv(paste(mydir, paste(sample, "CNVs.csv"), sep =""))

# output just the bed file
# remove chr 
Case_WGS$Position..hg38.<-gsub("chr","",as.character(Case_WGS$Position..hg38.))
Case_WGS_location <- separate(data = Case_WGS, col = Position..hg38., into = c("chr", "start", "end"))
Case_WGS_location_bed <- Case_WGS_location[,c("chr", "start", "end")]

# read in samples
total_files =  list.files(path = paste0(mydir,subfolder, sep= "" ), pattern= ("*.summarised_calls.tsv"))
sum_df  <- as.data.frame(matrix(ncol = 6, nrow = 0))
for(i in 1:length(total_files)) { #up to 4 samples
  file <- total_files[i]
  file_name <- strsplit(file, ".summarised_calls.tsv") #removes the .tsv taken from file name
  tsv_file <- read.csv(paste(mydir, subfolder, "/", file, sep= ""), sep="\t", header = T) #read in the file as a dataframe
  sum_df <- rbind(sum_df, tsv_file)
  file_name <- file_name[[1]] #get the dataframe from the list
  file_name <- gsub("_S.*", "", file_name)
  assign(file_name, tsv_file) #assigns the tsv_file to the object file_name which is the dataframe
  file_name <- NULL
  tsv_file <- NULL
}
sWGS_vs_WGS(Case_WGS, metadata, sample, mydir, subfolder, attributes, filters, mart)
```


### Downsampled 50kbp

#### How many TP, TN and FP when comparing our predicted CNs to the truth dataset

There will a tsv file that contains the truth bed and whether each segment in the WGS exists in the sWGS, this will be denotated with a "yes" in the column. If blank, it means that it is not called in the sWGS. Also, a table will be outputted detailed the TP, FP and TN.

* TP = Predicted CN in truth dataset
* FN = Not predicted but in truth set (e.g chrX and chrY as qDNASeq does autosomal chromosomes)
* FP = Predicted but not in truth set (keep in mind that it can be germline as we dont have control)
* TN = Not predicted in CN and not in truth dataset (can't be demonstrated)


```{r}
sample="CUH189"
# read  data in
mydir = "~/Documents/Projects/cancer/sWGS/"
subfolder = "case4_down_50bin"

# read in truth data
Case_WGS <- read.csv(paste(mydir, paste(sample, "CNVs.csv"), sep =""))

# output just the bed file
# remove chr 
Case_WGS$Position..hg38.<-gsub("chr","",as.character(Case_WGS$Position..hg38.))
Case_WGS_location <- separate(data = Case_WGS, col = Position..hg38., into = c("chr", "start", "end"))
Case_WGS_location_bed <- Case_WGS_location[,c("chr", "start", "end")]

# read in samples
total_files =  list.files(path = paste0(mydir,subfolder, sep= "" ), pattern= ("*.summarised_calls.tsv"))
sum_df  <- as.data.frame(matrix(ncol = 6, nrow = 0))
for(i in 1:length(total_files)) { #up to 4 samples
  file <- total_files[i]
  file_name <- strsplit(file, ".summarised_calls.tsv") #removes the .tsv taken from file name
  tsv_file <- read.csv(paste(mydir, subfolder, "/", file, sep= ""), sep="\t", header = T) #read in the file as a dataframe
  sum_df <- rbind(sum_df, tsv_file)
  file_name <- file_name[[1]] #get the dataframe from the list
  file_name <- gsub("_S.*", "", file_name)
  assign(file_name, tsv_file) #assigns the tsv_file to the object file_name which is the dataframe
  file_name <- NULL
  tsv_file <- NULL
}
sWGS_vs_WGS(Case_WGS, metadata, sample, mydir, subfolder, attributes, filters, mart)
```



### Downsampled 100kbp

#### How many TP, TN and FP when comparing our predicted CNs to the truth dataset

There will a tsv file that contains the truth bed and whether each segment in the WGS exists in the sWGS, this will be denotated with a "yes" in the column. If blank, it means that it is not called in the sWGS. Also, a table will be outputted detailed the TP, FP and TN.

* TP = Predicted CN in truth dataset
* FN = Not predicted but in truth set (e.g chrX and chrY as qDNASeq does autosomal chromosomes)
* FP = Predicted but not in truth set (keep in mind that it can be germline as we dont have control)
* TN = Not predicted in CN and not in truth dataset (can't be demonstrated)


```{r}
sample="CUH189"
# read  data in
mydir = "~/Documents/Projects/cancer/sWGS/"
subfolder = "case4_down_100bin"

# read in truth data
Case_WGS <- read.csv(paste(mydir, paste(sample, "CNVs.csv"), sep =""))

# output just the bed file
# remove chr 
Case_WGS$Position..hg38.<-gsub("chr","",as.character(Case_WGS$Position..hg38.))
Case_WGS_location <- separate(data = Case_WGS, col = Position..hg38., into = c("chr", "start", "end"))
Case_WGS_location_bed <- Case_WGS_location[,c("chr", "start", "end")]

# read in samples
total_files =  list.files(path = paste0(mydir,subfolder, sep= "" ), pattern= ("*.summarised_calls.tsv"))
sum_df  <- as.data.frame(matrix(ncol = 6, nrow = 0))
for(i in 1:length(total_files)) { #up to 4 samples
  file <- total_files[i]
  file_name <- strsplit(file, ".summarised_calls.tsv") #removes the .tsv taken from file name
  tsv_file <- read.csv(paste(mydir, subfolder, "/", file, sep= ""), sep="\t", header = T) #read in the file as a dataframe
  sum_df <- rbind(sum_df, tsv_file)
  file_name <- file_name[[1]] #get the dataframe from the list
  file_name <- gsub("_S.*", "", file_name)
  assign(file_name, tsv_file) #assigns the tsv_file to the object file_name which is the dataframe
  file_name <- NULL
  tsv_file <- NULL
}
sWGS_vs_WGS(Case_WGS, metadata, sample, mydir, subfolder, attributes, filters, mart)
```


### Downsampled 500kbp

#### How many TP, TN and FP when comparing our predicted CNs to the truth dataset

There will a tsv file that contains the truth bed and whether each segment in the WGS exists in the sWGS, this will be denotated with a "yes" in the column. If blank, it means that it is not called in the sWGS. Also, a table will be outputted detailed the TP, FP and TN.

* TP = Predicted CN in truth dataset
* FN = Not predicted but in truth set (e.g chrX and chrY as qDNASeq does autosomal chromosomes)
* FP = Predicted but not in truth set (keep in mind that it can be germline as we dont have control)
* TN = Not predicted in CN and not in truth dataset (can't be demonstrated)


```{r}
sample="CUH189"
# read  data in
mydir = "~/Documents/Projects/cancer/sWGS/"
subfolder = "case4_down_500bin"

# read in truth data
Case_WGS <- read.csv(paste(mydir, paste(sample, "CNVs.csv"), sep =""))

# output just the bed file
# remove chr 
Case_WGS$Position..hg38.<-gsub("chr","",as.character(Case_WGS$Position..hg38.))
Case_WGS_location <- separate(data = Case_WGS, col = Position..hg38., into = c("chr", "start", "end"))
Case_WGS_location_bed <- Case_WGS_location[,c("chr", "start", "end")]

# read in samples
total_files =  list.files(path = paste0(mydir,subfolder, sep= "" ), pattern= ("*.summarised_calls.tsv"))
sum_df  <- as.data.frame(matrix(ncol = 6, nrow = 0))
for(i in 1:length(total_files)) { #up to 4 samples
  file <- total_files[i]
  file_name <- strsplit(file, ".summarised_calls.tsv") #removes the .tsv taken from file name
  tsv_file <- read.csv(paste(mydir, subfolder, "/", file, sep= ""), sep="\t", header = T) #read in the file as a dataframe
  sum_df <- rbind(sum_df, tsv_file)
  file_name <- file_name[[1]] #get the dataframe from the list
  file_name <- gsub("_S.*", "", file_name)
  assign(file_name, tsv_file) #assigns the tsv_file to the object file_name which is the dataframe
  file_name <- NULL
  tsv_file <- NULL
}
sWGS_vs_WGS(Case_WGS, metadata, sample, mydir, subfolder, attributes, filters, mart)
```

#### FP, TP, FN rates across different bins

Different bins results in different TP, FP and FN rates. We can summarise this into one plot.

```{r}
sample="CUH189"
mydir = "~/Documents/Projects/cancer/sWGS/"
subfolder = "case4_down_500bin"
Case4_down_500bin = read.csv(paste(mydir,"/", subfolder, "/", sample ,"_TP_FP_TN.tsv", sep =""), sep="\t", header = T)
rownames(Case4_down_500bin) <- paste(rownames(Case4_down_500bin), subfolder, sep = "_")

subfolder = "case4_down_100bin"
Case4_down_100bin = read.csv(paste(mydir,"/", subfolder, "/", sample ,"_TP_FP_TN.tsv", sep =""), sep="\t", header = T)
rownames(Case4_down_100bin) <- paste(rownames(Case4_down_100bin), subfolder, sep = "_")

subfolder = "case4_down_50bin"
Case4_down_50bin = read.csv(paste(mydir,"/", subfolder, "/", sample ,"_TP_FP_TN.tsv", sep =""), sep="\t", header = T)
rownames(Case4_down_50bin) <- paste(rownames(Case4_down_50bin), subfolder, sep = "_")

subfolder = "case4_down_15bin"
Case4_down_15bin = read.csv(paste(mydir,"/", subfolder, "/", sample ,"_TP_FP_TN.tsv", sep =""), sep="\t", header = T)
rownames(Case4_down_15bin) <- paste(rownames(Case4_down_15bin), subfolder, sep = "_")

subfolder = "case4_down_5bin"
Case4_down_5bin = read.csv(paste(mydir,"/", subfolder, "/", sample ,"_TP_FP_TN.tsv", sep =""), sep="\t", header = T)
rownames(Case4_down_5bin) <- paste(rownames(Case4_down_5bin), subfolder, sep = "_")

Case4 <- rbind(Case4_down_5bin, Case4_down_15bin, Case4_down_50bin, Case4_down_100bin, Case4_down_500bin)
Case4$info <- rownames(Case4)
Case4 <- separate(data = Case4, col = info, sep = "_", into = c("SWGS", "sample", "sample_tissue",
                                                     "tissue_type", "case", "downsampled", "bin"))
Case4$tissue_type <- gsub('\\s+', '', Case4$tissue_type)
Case4$bin <- as.factor(gsub('bin*.', '', Case4$bin))
Case4$bin <- factor(gsub('bin*.', '', Case4$bin), levels=c(5,10,15,50,100,500))

Case4 <- separate(data = Case4, col = sample_tissue, sep = 4, into = c("ID", "repeats"))
#replace null with second
Case4$repeats <- gsub(' ', 'O', Case4$repeats)


Case4_FF <- Case4[which(Case4$tissue_type %in% "FF"),]
Case4_FFPE <- Case4[which(Case4$tissue_type %in% "FFPE"),]


ggplot(Case4_FF, aes(x = bin, y = TP, group = Case4_FF$tissu)) +
  geom_line(aes(color=Case4_FF$repeats)) +
  geom_point(aes(shape=Case4_FF$repeats))
```



```{r}
sessionInfo()
```

